src/main/scala/akka/persistence/jdbc/journal/JdbcAsyncWriteJournal.scala

```diff
@@ -98,7 +99,7 @@ class JdbcAsyncWriteJournal(config: Config) extends AsyncWriteJournal {
       .map(_ => ())
 
   override def postStop(): Unit = {
-    db.close()
+    db.close() // TODO maybe not?
```

Comment:
I find this tricky. I wasn't aware of that before I faced the issue related with JNDI `DataSource` and `AsyncExecutor`.

The problem that I see with this, and probably also the reason why you added this TODO, is that each plugin will be shutting down at different moments, but if they are sharing the same DB we may be closing it too soon.

I checked how the `akka-persistence-cassandra` is dealing with that and they simply have an independent cassandra session per plugin component.

As of today, the `akka-persistence-jdbc` is not sharing the db, so closing it independently is not an issue. Except that, in Lagom, we share the DB via JNDI. In that case, this is a problem for Lagom, but not for general usage of this plugin.

I think we must have a more robust way of sharing the database (or not) and at the same time, make sure they are all closed in a proper fashion. 

Maybe we should control it through `DefaultStickDatabaseProvider`.  We must somehow make it aware of each DB instance is being used by each plugin (journal, snapshot and read-journal). If they are being shared or not, etc. 

The close method should be delegated to `DefaultStickDatabaseProvider`. So, for example, if it knows that a DB is being shared by the journal and the read-journal, but not by the snapshot store, the first close call will have no effect, while the second will effectively close the shared DB.

Reply:
I think we should make the shutdown strategy configurable. The following configurations would be options:
- Close when the Journal or SnapshotJournal is closed (not possible for the ReadJournal). I agree that in case of database-sharing, we should have a mechanism for closing the db which knows if a database is used elsewhere.
- Close when the ActorSystem is closed (using CoordinatedShutdown from Akka 2.5), in this case we do not need to know if the database is used elsewhere.
- Do not close the db at all (allowing the user to do something manually)

Reply:
Maybe we should opt for `CoordinatedShutdown` in Akka 2.5. and don't backport it to 2.4 branch.

The current impl in 2.4 is not that bad, because it's not sharing the DB. Unless it's being shared through JNDI.